### PR TITLE
Add flash message support for login errors

### PR DESCRIPTION
**Pull request by cadeParade**
If there is some problem with the login system on our end (between the time github pings our backend and the time the user object shows up in the frontend), this displays a flash message that persists across 1 full page reload, which happens after a session is invalidated.

flash-messages

### CODE REVIEW

**Comment by cadeParade** [comment]
@PixelJanitor plz to help with nice flash messages 🙏

File: app/services/session.js [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
-import {resolve} from 'rsvp';
+import {resolve, reject} from 'rsvp';
 import {Promise as EmberPromise} from 'rsvp';
+import localStorageProxy from 'percy-web/lib/localstorage';
 
 export default SessionService.extend({
   store: service(),
   analytics: service(),
-
+  flashMessages: service(),
```
Comment (cadeParade) [inline app/services/session.js]: get rid of this

File: app/services/session.js [hunk 2]
```diff
@@ -52,10 +54,8 @@ export default SessionService.extend({
   _clearThirdPartyUserContext() {
     this._clearSentry();
     this._clearAnalytics();
-
-    if (window.localStorage) {
-      window.localStorage.clear();
```
Comment (cadeParade) [inline app/services/session.js]: This was clearing all of localstorage which doesn't work if we need to persist flash message data in localstorage. So I've changed this to remove specific keys with strings related to their services.

File: app/styles/_flash_messages.scss [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/* stylelint-disable selector-no-qualifying-type */
+
+.flash-message {
```
Comment (djones) [inline app/styles/_flash_messages.scss]: Let's run this by @PixelJanitor. We have some existing "alert" styles here https://github.com/percy/percy-web/blob/master/app/styles/_alerts.scss that would be a good starting point but that might not be visually acceptable in a flash message context.

File: app/lib/localstorage.js [hunk 4]
```diff
@@ -27,4 +27,29 @@ export default {
       // Safari throws errors while accessing localStorage in private mode.
     }
   },
+
+  keys() {
+    try {
+      // const localStorageLength = localStorage.length;
```
[the remark this answers is not in the document]
Comment (cadeParade) [reply app/lib/localstorage.js]: yes i could, thanks for catching

File: app/routes/application.js [hunk 5]
```diff
@@ -38,7 +39,20 @@ export default Route.extend(ApplicationRouteMixin, EnsureStatefulLogin, {
   },
 
   _loadCurrentUser() {
-    return this.get('session').loadCurrentUser();
+    return this.get('session')
+      .loadCurrentUser()
+      .catch(() => {
+        this.get('flashMessages').createPersistentFlashMessage(
+          {
```
[the remark this answers is not in the document]
Comment (cadeParade) [reply app/routes/application.js]: Any thoughts about wording?

**Comment by cadeParade** [comment]
@djones While writing up steps for Derek to reproduce the error, I realized it should show a flash message if a user runs into the same problem, so I've added a persistent flash message to the auth-failure route.

**Comment by cadeParade** [comment]
@PixelJanitor to get a flash message you can:
1. put `this.get('flashMessages').success('message', {sticky: true});` inside the `beforeModel` hook in `app/routes/application`. This will always show a message whenever the page loads. You can change the `success` to `danger`, `info`, or `warning` and it will change the class. You can also get more than one to display on the page at once, if you edit the absolute position css.

2. To reproduce the actual error state, start your server with export `PERCY_WEB_API_HOST=https://percy.io` and _not_ the token, and try to login. It should display a flash message.

The flash messages are from `https://github.com/poteto/ember-cli-flash`, so feel free to look at other settings to make it nicer. You can attach animations to it! I tried briefly, but had no luck.

File: app/styles/_flash_messages.scss (hunk 3 above)
Comment (PixelJanitor) [reply app/styles/_flash_messages.scss]: are those alert styles used in a current alert somewhere?

Comment (djones) [reply app/styles/_flash_messages.scss]: Yep, there's a couple of samples here https://percy.io/docs/clients/javascript/react

**Comment by cadeParade** [comment]
Omg these are very better 😍

**Comment by PixelJanitor** [comment]
This is good to go on my end if you want to do a final review 👍

File: app/routes/auth-failure.js [hunk 6]
```diff
@@ -1,6 +1,19 @@
...
+  flashMessages: service(),
   beforeModel() {
+    this.get('flashMessages').createPersistentFlashMessage(
+      {
+        type: 'danger',
+        message:
+          'There was a problem with logging in. \
+        Please try again or contact us if the issue does not resolve.',
+      },
+      {persistentReloads: 0},
```
Comment (cadeParade) [inline app/routes/auth-failure.js]: me: double check this

**Comment by cadeParade** [comment]
@PixelJanitor The flash messages sort of disappear on mobile screen sizes -- they disappear very gracefully, but wondering if they should still display on mobile?

**Comment by PixelJanitor** [comment]
My mobile support commit didn't make it for some reason, but I added it back in and added a small refactor to a sass loop. Let me know if you need anything else on this!

**Comment by cadeParade** [comment]
Thanks @PixelJanitor ! I've confirmed the fix and merging now!

**Pull request merged**